Bad file descriptor

More often than not, I'm getting the following exception while training (e.g., training on MSR-VTT):

```
Exception in thread QueueFeederThread:
Traceback (most recent call last):
  File "***/miniconda3/envs/frozen/lib/python3.7/multiprocessing/queues.py", line 232, in _feed
    close()
  File "***/miniconda3/envs/frozen/lib/python3.7/multiprocessing/connection.py", line 177, in close
    self._close()
  File "***/miniconda3/envs/frozen/lib/python3.7/multiprocessing/connection.py", line 361, in _close
    _close(self._handle)
OSError: [Errno 9] Bad file descriptor

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "***/miniconda3/envs/frozen/lib/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "***/miniconda3/envs/frozen/lib/python3.7/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "***/miniconda3/envs/frozen/lib/python3.7/multiprocessing/queues.py", line 263, in _feed
    queue_sem.release()
ValueError: semaphore or lock released too many times
```

Do you have any idea of what's going on?

The training still continues and finishes (reporting repeated errors), not sure if it trains fine though.

Just to give more details, after that I get similar "Bad file descriptor" errors but also the following one repeated times:

```
Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7f3c8486f7a0>
Traceback (most recent call last):
  File "***/miniconda3/envs/frozen/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 1324, in __del__
    self._shutdown_workers()
  File "***/miniconda3/envs/frozen/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 1316, in _shutdown_workers
    if w.is_alive():
  File "***/miniconda3/envs/frozen/lib/python3.7/multiprocessing/process.py", line 151, in is_alive
    assert self._parent_pid == os.getpid(), 'can only test a child process'
AssertionError: can only test a child process
```

--- a/base/base_dataset.py
+++ b/base/base_dataset.py
@@ -2,6 +2,8 @@ import random
 import cv2
 import av
 import os
+
+import decord
 import numpy as np
 import torch
 from PIL import Image, ImageFile
@@ -296,6 +298,19 @@ def read_frames_av(video_path, num_frames, sample='rand', fix_start=None):
     return frames, frame_idxs
 
 
+decord.bridge.set_bridge("torch")
+
+
+def read_frames_decord(video_path, num_frames, sample='rand', fix_start=None):
+    video_reader = decord.VideoReader(video_path, num_threads=1)
+    vlen = len(video_reader)
+    frame_idxs = sample_frames(num_frames, vlen, sample=sample, fix_start=fix_start)
+    frames = video_reader.get_batch(frame_idxs)
+    frames = frames.float() / 255
+    frames = frames.permute(0, 3, 1, 2)
+    return frames, frame_idxs
+
+
 def get_video_len(video_path):
     cap = cv2.VideoCapture(video_path)
     if not (cap.isOpened()):
@@ -307,5 +322,6 @@ def get_video_len(video_path):
 
 video_reader = {
     'av': read_frames_av,
-    'cv2': read_frames_cv2
+    'cv2': read_frames_cv2,
+    'decord': read_frames_decord
 }

--- a/requirements/frozen.yml
+++ b/requirements/frozen.yml
@@ -129,6 +129,7 @@ dependencies:
     - bravado-core==5.17.0
     - colorama==0.4.4
     - cycler==0.10.0
+    - decord==0.6.0
     - dominate==2.6.0
     - einops==0.3.0
     - future==0.18.2

